File: services/submission_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"doctor-service/dto"
	"doctor-service/entities"
	"doctor-service/repositories"

	"github.com/google/uuid"
)

type SubmissionService interface {
	SubmitVerification(ctx context.Context, doctorPublicID uuid.UUID) (*dto.SubmitVerificationResponse, error)
}

type DefaultSubmissionService struct {
	doctorRepo  repositories.DoctorRepository
	licenseRepo repositories.LicenseRepository
	qualRepo    repositories.QualificationRepository
	clinicRepo  repositories.ClinicRepository
	docRepo     repositories.DocumentRepository
	historyRepo repositories.VerificationHistoryRepository
	jobRepo     repositories.JobRepository
}

func NewSubmissionService(
	doctorRepo repositories.DoctorRepository,
	licenseRepo repositories.LicenseRepository,
	qualRepo repositories.QualificationRepository,
	clinicRepo repositories.ClinicRepository,
	docRepo repositories.DocumentRepository,
	historyRepo repositories.VerificationHistoryRepository,
	jobRepo repositories.JobRepository,
) SubmissionService {
	return &DefaultSubmissionService{
		doctorRepo:  doctorRepo,
		licenseRepo: licenseRepo,
		qualRepo:    qualRepo,
		clinicRepo:  clinicRepo,
		docRepo:     docRepo,
		historyRepo: historyRepo,
		jobRepo:     jobRepo,
	}
}

func (s *DefaultSubmissionService) SubmitVerification(ctx context.Context, doctorPublicID uuid.UUID) (*dto.SubmitVerificationResponse, error) {
	// 1. Resolve Doctor
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor account not found")
	}

	var missingChecklist []string

	// Checklist 1: Mobile Verified
	if !doc.MobileVerified {
		missingChecklist = append(missingChecklist, "Mobile number verification pending")
	}

	// Checklist 2: Medical License
	licenses, err := s.licenseRepo.FindByDoctorID(ctx, doc.ID)
	if err != nil || len(licenses) == 0 {
		missingChecklist = append(missingChecklist, "At least one medical registration license is required")
	}

	// Checklist 3: Qualifications
	quals, err := s.qualRepo.FindByDoctorID(ctx, doc.ID)
	if err != nil || len(quals) == 0 {
		missingChecklist = append(missingChecklist, "At least one medical qualification degree is required")
	}

	// Checklist 4: Clinics
	clinics, err := s.clinicRepo.FindByDoctorID(ctx, doc.ID)
	if err != nil || len(clinics) == 0 {
		missingChecklist = append(missingChecklist, "At least one clinic listing is required")
	}

	// Checklist 5: Mandatory Documents
	uploadedDocs, err := s.docRepo.FindByDoctorID(ctx, doc.ID)
	if err != nil {
		uploadedDocs = []entities.DoctorDocument{}
	}

	hasRegistrationCert := false
	hasDegreeCert := false
	hasGovtID := false

	for _, d := range uploadedDocs {
		switch d.DocumentType {
		case entities.DocumentTypeRegistrationCertificate:
			hasRegistrationCert = true
		case entities.DocumentTypeMBBSCertificate, entities.DocumentTypeMDCertificate, entities.DocumentTypeMedicalDegreeCertificate:
			hasDegreeCert = true
		case entities.DocumentTypeAadhaar, entities.DocumentTypePAN, entities.DocumentTypePassport:
			hasGovtID = true
		}
	}

	if !hasRegistrationCert {
		missingChecklist = append(missingChecklist, "Medical Registration Certificate document is missing")
	}
	if !hasDegreeCert {
		missingChecklist = append(missingChecklist, "Medical Degree Certificate document is missing")
	}
	if !hasGovtID {
		missingChecklist = append(missingChecklist, "Government Identity proof (Aadhaar, PAN, or Passport) is missing")
	}

	if len(missingChecklist) > 0 {
		return nil, fmt.Errorf("verification submission rejected. Incomplete profile checklist:\n- %s", strings.Join(missingChecklist, "\n- "))
	}

	// 2. Transition Status to PENDING
	doc.Status = entities.DoctorStatusPending
	if err := s.doctorRepo.Update(ctx, doc); err != nil {
		return nil, fmt.Errorf("failed to update doctor status: %w", err)
	}

	// 3. Record Audit Log in Verification History
	remarks := "Doctor submitted profile and all mandatory documents for verification."
	history := &entities.VerificationHistory{
		DoctorID:    doc.ID,
		Action:      "SUBMISSION_COMPLETED",
		Status:      string(entities.DoctorStatusPending),
		Remarks:     &remarks,
		PerformedAt: time.Now(),
	}
	_ = s.historyRepo.Create(ctx, history)

	// 4. Enqueue Verification Background Job in QUEUED status
	if s.jobRepo != nil {
		job := &entities.VerificationJob{
			JobID:      uuid.New(),
			DoctorID:   doc.PublicID,
			JobType:    entities.JobTypeFullPipeline,
			Status:     entities.JobStatusQueued,
			Priority:   1,
			MaxRetries: 3,
		}
		_ = s.jobRepo.Create(ctx, job)
	}

	return &dto.SubmitVerificationResponse{
		Message:     "Doctor verification application submitted successfully. Your profile has been queued for background OCR and council verification.",
		PublicID:    doc.PublicID,
		Status:      string(doc.Status),
		SubmittedAt: time.Now(),
	}, nil
}

File: python_backend/services/submission_service.py
from datetime import datetime, timezone
from entities.doctor import DoctorStatus
from entities.history import VerificationHistory
from entities.job import VerificationJob, JobType, JobStatus


class SubmissionService:
    def SubmitVerification(self, ctx, doctor_public_id):
        pass


class DefaultSubmissionService(SubmissionService):
    def __init__(self, doctor_repo, license_repo, qual_repo, clinic_repo, doc_repo, history_repo, job_repo):
        self.doctorRepo = doctor_repo
        self.licenseRepo = license_repo
        self.qualRepo = qual_repo
        self.clinicRepo = clinic_repo
        self.docRepo = doc_repo
        self.historyRepo = history_repo
        self.jobRepo = job_repo

    def SubmitVerification(self, ctx, doctor_public_id):
        doc = self.doctorRepo.FindByPublicID(str(doctor_public_id))
        if not doc:
            raise Exception("doctor account not found")

        missing_checklist = []

        if not doc.mobile_verified:
            missing_checklist.append("Mobile number verification pending")

        licenses = self.licenseRepo.FindByDoctorID(doc.id) or []
        if not licenses:
            missing_checklist.append("At least one medical registration license is required")

        quals = self.qualRepo.FindByDoctorID(doc.id) or []
        if not quals:
            missing_checklist.append("At least one medical qualification degree is required")

        clinics = self.clinicRepo.FindByDoctorID(doc.id) or []
        if not clinics:
            missing_checklist.append("At least one clinic listing is required")

        uploaded_docs = self.docRepo.FindByDoctorID(doc.id) or []
        has_registration_cert = False
        has_degree_cert = False
        has_govt_id = False

        for d in uploaded_docs:
            dtype = d.document_type.value if hasattr(d.document_type, "value") else str(d.document_type)
            if dtype == "REGISTRATION_CERTIFICATE":
                has_registration_cert = True
            elif dtype in ("MBBS_CERTIFICATE", "MD_CERTIFICATE", "MEDICAL_DEGREE_CERTIFICATE"):
                has_degree_cert = True
            elif dtype in ("AADHAAR", "PAN", "PASSPORT"):
                has_govt_id = True

        if not has_registration_cert:
            missing_checklist.append("Medical Registration Certificate document is missing")
        if not has_degree_cert:
            missing_checklist.append("Medical Degree Certificate document is missing")
        if not has_govt_id:
            missing_checklist.append("Government Identity proof (Aadhaar, PAN, or Passport) is missing")

        if missing_checklist:
            missing_str = "\n- ".join(missing_checklist)
            raise Exception(f"verification submission rejected. Incomplete profile checklist:\n- {missing_str}")

        doc.status = DoctorStatus.PENDING
        try:
            self.doctorRepo.Update(doc)
        except Exception as e:
            raise Exception(f"failed to update doctor status: {e}")

        try:
            self.historyRepo.Create(
                VerificationHistory(
                    doctor_id=str(doc.id),
                    action="SUBMISSION_COMPLETED",
                    status="PENDING",
                    remarks="Doctor submitted profile and all mandatory documents for verification.",
                )
            )
        except Exception:
            pass

        if self.jobRepo:
            try:
                self.jobRepo.Create(
                    VerificationJob(
                        doctor_id=str(doc.id),
                        job_type=JobType.FULL_PIPELINE,
                        status=JobStatus.QUEUED,
                        priority=1,
                        max_retries=3,
                    )
                )
            except Exception:
                pass

        class SubmitVerificationResponse:
            pass

        resp = SubmitVerificationResponse()
        resp.message = (
            "Doctor verification application submitted successfully. "
            "Your profile has been queued for background OCR and council verification."
        )
        resp.public_id = str(doc.public_id)
        resp.status = doc.status.value if hasattr(doc.status, "value") else str(doc.status)
        resp.submitted_at = datetime.now(timezone.utc)
        return resp
